src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java

```diff
@@ -573,16 +573,22 @@ private static Long findSchema(Connection connection, Position targetPosition, L
 			// Only consider binlog positions before the target position on the current server.
 			// Within those, sort for the latest binlog file, then the latest binlog position.
 			PreparedStatement s = connection.prepareStatement(
-				"SELECT id from `schemas` "
-				+ "WHERE deleted = 0 "
-				+ "AND last_heartbeat_read <= ? AND ((binlog_file < ?) OR (binlog_file = ? and binlog_position < ?)) AND server_id = ? "
-				+ "ORDER BY last_heartbeat_read DESC, binlog_file DESC, binlog_position DESC limit 1");
+				"SELECT id from `schemas` " +
+				"WHERE deleted = 0 AND " +
+				"last_heartbeat_read <= ? AND " +
+				"((binlog_file < ?) OR " +
+				"(binlog_file = ? AND binlog_position < ?) OR " +
+				"(binlog_file = ? AND binlog_position = ? AND deltas = '[]')) AND " + // master failover
```

Comment:
Why can't we just make this `binlog_position <=` instead of one condition for `<` and one for `=` which only matches empty deltas? I'm assuming there's a good reason, but I'd like to remember it ;)

Reply:
@timbertson has an interesting idea here, but I think "equal to our position" doesn't quite make sense.  better would be to keep the last-applied delta in-memory and check that we haven't already applied it.

Reply:
How would you check that, do you have to ask mysql about the changes and see if they match the before or after state of the delta you're talking about? That sounds complex :(

Reply:
nah, we suck all the deltas into memory when we boot: https://github.com/zendesk/maxwell/blob/master/src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java#L387.  All we'd *really* need to do is to stash away the position of the last delta we processed.  If we hit a schema change at a position that matches the stashed position (as it would in the dupe-schema case), we can ignore it.

Reply:
Maybe we could clarify the scenarios here:

 1. Failover and then crash. We'll have a `position` equal to the last schema, which was a "no-op" schema inserted when we discovered the corresponding position from our old master. It doesn't matter whether we apply this delta (it's empty), but we must apply either this schema or its base_schema_id, not an old schema for the current server_id.
2. Schema change then crash. We'll store the DDL position in `positions`, but when we restart we'll begin "at" the DDL and will process it again. We want to pretend the most recent delta didn't happen, so that we can reprocess it without crashing. This could be done either by selecting a strictly-less-than position, or by ignoring the final delta in the chain when doing the restore.
3. Any position which is between schema changes. We always want to apply the full schema chain, since the DDL has definitely been applied.

It seems like maybe we're storing `position` inconsistently. For the case of regular data, we use `RowMap.getPosition()`, which is actually `nextPosition`. So the stored position is "the next message we should consume".
Whereas in the case of DDL, we're apparently processing a DDL message, then setting our position to the _beginning_ of that DDL, a.k.a "the last message we consumed". Shouldn't we be saving a position pointing to the message _after_ that DDL. I don't know where this code path is, do you @osheroff ?

If we are storing positions inconsistently, we should just fix that instead and then this problem goes away 🤞 .

Reply:
To answer my own question: I think it is consistent, but the schema we're storing in the case of a failover is not valid. That schema's position is pointing to "the next message you'll process", but the schema it represents is in the past.

The hacky thing would be to take `1` off the position when we store a recovery schema so that it's strictly in the past, but that's pretty gross because it's not a valid position (does it matter though? it's a fake schema anyway...). A proper solution would return the next `position` when doing a recovery, but would create a schema for the "last" position (i.e. the previous event).

Reply:
To be clear, I guess this pull request won't fix the scenario where binlog position is pointing to DDL and master failover happen at the same time. It would still process DDL twice. 

So can we only insert position table if it is DML ?